File: PDP/Odev2/KISIM1/Odev2Kutuphane/src/RandomPacket/IMEINo.java
/**
*
* @author Melih Ensar BARIŞIK ensar.barisik@ogr.sakarya.edu.tr / Ömer Can ÇALIŞIR omer.calisir@ogr.sakarya.edu.tr
* @since  25.04.2020
* <p>
*   İmei numarası oluşturulan kısım
* </p>
*/
package RandomPacket;

/**
 *
 * @author bar_i
 */
public class IMEINo {
    
    
    private String IMEIS;
    
    public IMEINo()
    {
        
        IMEIS = "";
    }
    
     public String getIMEINo() {
        return IMEIS;
    }
     
     //Imei nuamrasının ilk on dört hanesini rastgele oluşturan fonksiyon
     public void constituteImeıNoPart1()
    {
        //Imei numarasının 14 rakamını random şekilde veriyoruz.
        IMEIS+= Rastgele.produceRandomNumber(14);
        
    }
     //verilen sayının rakamlarını toplama işlemi
     public static int collectNumbers(int data)
    {
        int total =0;
        
        while(data != 0)
        {
            // Gelen datanın basamaklarını toplamak için on ile modunu alıp son karakteri toplama ekliyoruz. 
            total+=(data%10);
            //Sayıyı on ile bölüp bir sonraki rakama geçiyoruz.
            data = data/10;
        }
        
       return total;
      
        
    }
     
     //İmei numarasının son hanesi hesaplanıyor.
    public  void calculateRecentDigit()
    {
        int control =0;
        int sum = 0;
       
        for(int i = 0; i<IMEIS.length(); i++)
        {
            if(i%2 ==0)
            {
                //Tek basamağa denk gelen haneler toplanıyor.
            int value = Character.getNumericValue(IMEIS.charAt(i));
            sum+=value;      
            }

            else
            {   //Çift basamağa denk gelen hanelerin ise iki ile çarpılıp oluşan sayının basamakları toplanıyor.
                int value = Character.getNumericValue(IMEIS.charAt(i));
                sum+=collectNumbers(value*2);
            }
        }
           
        // son hanenin atanması işlemi (son mod alma parantez içinin 10 çıkması olasılığına karşı yapıldı)
        String kontrolHanesi = String.valueOf((10-(sum % 10 ))%10);
        IMEIS+=kontrolHanesi;
       
        
  
    }
    
    //Oluşturma fonksiyonları tek tek çağırılarak işlemi sıralı yapamya yarıyor.
    public void createImei()
    {
        
       constituteImeıNoPart1();
       calculateRecentDigit();
      
    }
    
    //Verilen imei numarasını kontrol ediyor.
    public static boolean controlIMEI(String kontrolEdilecekIMEINo)
    {
        int toplam = 0;
       
        //İlk 14 hanenin rakamlarını toplama işlemini yapıyor.
        for (int i = 0; i < kontrolEdilecekIMEINo.length() - 1; i++) {
            if (i % 2 == 0) {
                //Tek basamağa denk gelen rakamalar toplanıyor.
                toplam+=Character.getNumericValue(kontrolEdilecekIMEINo.charAt(i));
            } else {
                 //Çift basamağa denk gelen rakamlar ise iki le çarpılıp oluşan sayının basamakları toplanıyor.
                 toplam+=collectNumbers(Character.getNumericValue(kontrolEdilecekIMEINo.charAt(i))*2);
            }
            
    }
         // son hanenin atanması işlemi (son mod alma parantez içinin 10 çıkması olasılığına karşı yapıldı)
        String kontrolHanesi = String.valueOf((10 - (toplam % 10)) % 10);
       
        //İmei numarası 15 rakamdan oluşmalı bunu kontrol ediyor.
        if(kontrolEdilecekIMEINo.length() != 15)
        {
            return false;
        }
        //Son hane olan kontrol hanesinni doğru hesaplanıp hesaplanmadğını kontrol ediyor.
        if (kontrolHanesi.equals(String.valueOf(kontrolEdilecekIMEINo.charAt(14)))){
            return true;
        }
        
        return false;
         }
   }
    
    

File: PDP/Odev2/KISIM2/src/IMEINo.c
/**
*
* @author Melih Ensar BARIŞIK ensar.barisik@ogr.sakarya.edu.tr / Ömer Can ÇALIŞIR omer.calisir@ogr.sakarya.edu.tr
* @since  25.04.2020
* <p>
*   İmei numarası oluşturulan kısım
* </p>
*/
#include "IMEINo.h"
#include "Rastgele.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include<string.h>


// nesne yönelimli benzetme için sözde kurucu fonksiyonu
IMEINo IMEINoUret()
{
	IMEINo this;
    // struct için bellekte yer ayırma işlemi
	this=(IMEINo)malloc(sizeof(struct IMEINO));

	
	int i;
	
    // ilk değer ataması yapma işlemi
	for(i=0; i<this->length; i++) {this->IMEIS[i] =0;}
	

	// fonksiyon göstericilerine oluşturulan
    // fonksiyonların adreslerinin verilmesi işlemi
	this->calculateRecentDigit=&_calculateRecentDigit;
	this->assignRandomDigits=&_assignRandomDigits;
	this->createImeiNo=&_createImeiNo;
	this->getImeiNo = &_getImeiNo;
	this->yokediciIMEINo=&_yokediciIMEINo;


	return this;
}


// imeinonun ilk 14 hanesinin random oluşturulması işlemini gerçekleyen fonksiyon
void _assignRandomDigits(const IMEINo this, Rastgele R)
{
	
	
	int i;
	for(i=0; i<14; i++)
	{
		// üretilen random rakamların atanması işlemi
		this->IMEIS[i]= R->randomRakamUret(R,0,10);
	}
}

// verilen sayının basamakları toplamını döndürmeye yarayan fonksiyon
int _collectNumbers(int data)
    {
        int total =0;
        
        while(data != 0)
        {   
            // ilk basamağı toplama ekleyip bir sonraki basamak için sayıyı küçültme işlemi
            total+=(data%10);
            data = data/10;
        }
        
       return total;
      
        
    }
 // imei numarasının son basamağını algoritmaya uygun olarak
 // hesaplama işlemini gerçekleyen fonksiyon
 void _calculateRecentDigit(const IMEINo this)
 {
	  int control =0;
      int sum = 0;
       
		int i;
        // imei numarasının ilk 14 hanesinde dolaşmaya yarayan for döngüsü
        for(i = 0; i<14; i++)
        {
			
			
            if(i%2 ==0)
            {
            // imei numarasındaki tek basamakların toplanması işlemi
            int value = this->IMEIS[i];
            sum+=value;      
            }

            else
            {   
                // imei numarasındaki çift basamakların 2 ile çarpılarak toplanması işlemi
                int value = this->IMEIS[i];
                sum+=_collectNumbers(value*2);
            }
        }
        
	   // son hanenin atanması işlemi (son mod alma parantez içinin 10 çıkması olasılığına karşı yapıldı)
		this->IMEIS[14]=(10-(sum % 10 ))%10;
	
       
 }
 
// imeino yu oluşturan fonksiyonların senkronize çalışmasını sağlayan fonksiyon
void _createImeiNo(const IMEINo this, Rastgele R){
	this->assignRandomDigits(this,R);
	this->calculateRecentDigit(this);
}

// imeino dizisini döndüren get fonksiyonu
int* _getImeiNo(const IMEINo this)
{
	return this->IMEIS;
}
 
// imeino nun doğruluğunu kontrol eden fonksiyon
bool controlIMEI(int *kontrolEdilecekNumara,int uzunluk)
 {
         // imei numarasının 15 haneli olup olmadığının kontrolü
         if(uzunluk != 15)
        {
            return false;
        }


        // basamakların kurala uygun olarak toplamının bulunması işlemleri
        int toplam = 0;
        for (int i = 0; i < uzunluk - 1; i++)
        {
            if (i % 2 == 0) {
                toplam+=kontrolEdilecekNumara[i];
            } else {

                 toplam+=_collectNumbers(kontrolEdilecekNumara[i]*2);
            }
        }
        


        // kontrol hanesinin bulunmasına yarayan matematik işlemi
        int kontrolHanesi = (10 - (toplam % 10)) % 10;

       
       // son hanenin kontrolünün yapılması işlemi
        if (kontrolHanesi==kontrolEdilecekNumara[14]){
            return true;
        }

        return false; 
     
    }

// bellekten alınan yerin geri teslimi için yıkıcı fonksiyon
void _yokediciIMEINo(IMEINo this)
{
	if(this==NULL);
	free(this);
}	